source/base/DateTime.ooc

```diff
@@ -14,22 +14,263 @@
+		if (This dateIsValid(year, month, day))
+			this _ticks = This dateToTicks(year, month, day)
+		else
+			raise ("invalid input specified for constructor(year,month,day)")
+	}
+	init: func@ ~fromHourMinuteSec (hour, minute, second, millisecond: Int) {
+		if (This timeIsValid(hour, minute, second, millisecond))
+			this _ticks = This timeToTicks(hour, minute, second, millisecond)
+		else
+			raise ("invalid input specified for constructor(hour,minute,second)")
+	}
+	init: func@ ~fromDateTime (year, month, day, hour, minute, second, millisecond: Int) {
+		if (This dateIsValid(year, month, day) && This timeIsValid(hour, minute, second, millisecond))
+			this _ticks = This dateToTicks(year, month, day) + This timeToTicks(hour, minute, second, millisecond)
+		else
+			raise ("invalid input specified for constructor(year,month,day,hour,minute,second,ms)")
+	}
 	isLeapYear: static func (year: Int) -> Bool { (year % 100 == 0) ? (year % 400 == 0) : (year % 4 == 0) }
 	kean_base_dateTime_getTicks: unmangled func -> UInt64 { this _ticks }
 	new: unmangled(kean_base_dateTime_new) static func ~API (ticks: UInt64) -> This { This new(ticks) }
 
+	millisecond: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) millisecond
+	}
+	second: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) second
+	}
+	minute: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) minute
+	}
+	hour: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) hour
+	}
+	day: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) day
+	}
+	month: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) month
+	}
+	year: func -> Int {
+		This _ticksToDateTimeHelper(this ticks) year
+	}
+
+	toString: func -> String {
+		this toStringFormat(This defaultFormat)
+	}
+	// <summary>
+	// Convert this object to string representation
+	// </summary>
+	// supported formatting expressions:
+	//	%yyyy - year
+	//	%yy	- year as two digit number
+	//  %MM - month with leading zero
+	//	%M	- month without leading zero
+	//	%dd - day with leading zero
+	//	%d	- day without leading zero
+	//	%hh - hour
+	//	%mm - minute
+	//	%ss - second
+	//	%zzzz - millisecond
+	// <param name="format">output format specification</param>
+	toStringFormat: func (format: String) -> String {
+		result := format
+		data := This _ticksToDateTimeHelper(this ticks)
+		result = result replaceAll("%yyyy", "%d" format(data year))
+		result = result replaceAll("%yy", "%d" format(data year % 100))
+		result = result replaceAll("%MM", (data month < 10 ? "0%d" : "%d") format(data month))
+		result = result replaceAll("%dd", (data day < 10 ? "0%d" : "%d") format(data day))
+		result = result replaceAll("%M", "%d" format(data month))
+		result = result replaceAll("%d", "%d" format(data day))
+		result = result replaceAll("%hh", "%d" format(data hour))
+		result = result replaceAll("%mm", "%d" format(data minute))
+		result = result replaceAll("%ss", "%d" format(data second))
```

Comment:
Hour, date, and minute should always be written as 2 digits, imo. Use `"%02d"`, as I mentioned above (assuming it actually works!).